### PR TITLE
Add Real-Time Emotion Detector (EmotionIQ)

### PR DESCRIPTION
**Pull request by ChitteshKumar**
I have developed a real-time emotion detector, which I named **EmotionIQ**. This tool leverages a CNN model, OpenCV techniques, and Streamlit to successfully detect facial expressions and classify emotions in real-time.

**EmotionIQ Implementation:**
- I added a new folder to the models directory, named emotion_detector, which contains the necessary dataset and model files for EmotionIQ.
- The dataset used comes from Kaggle: [Face Expression Recognition Dataset](https://www.kaggle.com/datasets/jonathanoheix/face-expression-recognition-dataset). I used this dataset to train the CNN model using TensorFlow, with the model being implemented in a .ipynb file.
- The trained model is now used within the driver.py file, where it processes real-time video captured from the camera, detects faces using OpenCV, and classifies the emotions. Additionally, warnings are displayed to the user, notifying them when the camera is being accessed.

**Integration into the Platform:**
As most of the code in the project was geared towards form-based interactions and making predictive calculators, I introduced a new function called render_detector() in the page_handler.py file. This function allows seamless integration of emotion detectors (or similar applications) into the platform, enabling anyone to easily add this feature in the future.

Files Modified:
- models/emotion_detector: 
    - Contains the dataset and the trained CNN model.
    - Model is built in a TensorFlow-based .ipynb file.
- driver.py:
    - Loads and runs the model, utilizing OpenCV for facial detection and real-time camera feed for emotion analysis.
- page_handler.py:
    - Added the render_detector() function to allow integration of emotion detection functionality into the platform.

By merging this, the project will now support a fully-functional, real-time emotion detection system, expanding the predictive capabilities beyond just calculators.

**Expected Outcome:**
The EmotionIQ feature will enable real-time emotion detection using a CNN model, enhancing user engagement with live facial expression analysis via webcam. The new render_detector() function in page_handler.py allows seamless integration of real-time detection features, improving modularity and reusability for future development.

### CODE REVIEW

**Comment by ChitteshKumar** [comment]
Hi @yashasvini121, I have completed the assigned task #31 . And created a pull request, kindly check it and provide feedback.
Looking forward to get the request accepted

Thank you

**Review by yashasvini121 — CHANGES_REQUESTED** [review]
Please upload your data folder in a ZIP format, as we will only be using the saved model at the end.

**Comment by ChitteshKumar** [comment]
> Please upload your data folder in a ZIP format, as we will only be using the saved model at the end. (Please confirm this, as I haven't checked the code.)

okay, I'll put the dataset into zip. I a bit unsure how to do that, should I close the PR and raise it again? can you tell me the procedure.

**Comment by yashasvini121** [comment]
@ChitteshKumar, just continue working on this branch; your changes will be reflected automatically.

**Comment by ChitteshKumar** [comment]
> @ChitteshKumar, just continue working on this branch; your changes will be reflected automatically.
> 
> Eg: You recently merged `master` into this pr related branch, those changes are now showing in the PR, look at the timeline, above.

okay I have tried doing. Please check it and tell me if its correct or not.

**Comment by ChitteshKumar** [comment]
hi @yashasvini121 , I would just like to remind you to kindly check this merge and complete it.

File: models/emotion_detector/.DS_Store
Comment (yashasvini121) [inline models/emotion_detector/.DS_Store:1]: For my general knowledge, could you please explain the purpose of the `.DS_Store` file in this project? Or else, If it’s not necessary, kindly remove it.

File: models/.DS_Store
Comment (yashasvini121) [inline models/.DS_Store:1]: For my general knowledge, could you please explain the purpose of the `.DS_Store` file in this project? Or else, If it’s not necessary, kindly remove it.

**Review by yashasvini121 — CHANGES_REQUESTED** [review]
Look into this error:
image

P.S.: Your model mostly returns a surprise emotion.

File: models/.DS_Store
Comment (ChitteshKumar) [reply models/.DS_Store:1]: I'll remove it as it just contains stores custom attributes of its containing folder, such as folder view options, icon positions, and other visual information. It is automatically generated in Mac.

File: models/emotion_detector/driver.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+    st.write("Click on start to use the webcam and detect your face emotion.")
+    # Start webcam streamer
+    webrtc_streamer(key="example", mode=WebRtcMode.SENDRECV, rtc_configuration=RTC_CONFIGURATION, video_processor_factory=FaceEmotion)
+
+if __name__ == "__main__":
+    get_prediction()
+
+
+
+
```
[the remark this answers is not in the document]
Comment (ChitteshKumar) [reply models/emotion_detector/driver.py:70]: I want it to be there.

File: requirements.txt [hunk 2]
```diff
@@ -152,4 +152,12 @@ webcolors==24.8.0
 webencodings==0.5.1
 websocket-client==1.8.0
 widgetsnbextension==4.0.13
+xgboost
+#for emotion detector
+tensorflow
+opencv-python
+streamlit
```
[the remark this answers is not in the document]
Comment (ChitteshKumar) [reply requirements.txt:159]: okay I'll do it

File: models/emotion_detector/driver.py [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+import json 
+
+from tensorflow.keras.preprocessing import image
+from tensorflow.keras.preprocessing.image import img_to_array, load_img
+from tensorflow.keras.models import load_model
+
+import streamlit as st
+from streamlit_webrtc import webrtc_streamer, VideoTransformerBase, RTCConfiguration, VideoProcessorBase, WebRtcMode
+
+model = load_model('/Users/chittesh/GSSoC/predictive-calc/models/emotion_detector/saved_models/best_model.keras')
```
[the remark this answers is not in the document]
Comment (ChitteshKumar) [reply models/emotion_detector/driver.py:13]: Okay I will do it. Apologise for not changing that

**Comment by ChitteshKumar** [comment]
> Look into this error: image
> 
> P.S.: Your model mostly returns a surprise emotion.

thank you so much for pointing the error out. I have tried fixing the error, can you check. Also the model is bit simple, so once this gets merged (if you allow), then I can create a new issue and work on it. Looking forward to hear from you soon.

**Comment by ChitteshKumar** [comment]
@yashasvini121 , hi can you review this please accept this.

**Comment by yashasvini121** [comment]
Will be reviewed by today

**Comment by yashasvini121** [comment]
Sorrry for the delay, for now, please instead of `yashasvini121:master`, create a pr to push to `yashasvini121:emotion-detector` branch.

**Pull request closed without merging**